### PR TITLE
refactor(request): split fetch-helpers along its natural seams

### PR DESCRIPTION
**Pull request by ndycode**
## Summary

Splits `lib/request/fetch-helpers.ts` (1,497 lines, 30+ exports) along its natural seams into four focused modules, per audit roadmap §4.1.2 (`docs/audits/AUDIT_2026-06-10.md`, PR #522). All code moved verbatim; `fetch-helpers.ts` remains the re-export facade, so **no importer changes and zero behavior change**.

## Changes

| File | Lines | Contents |
| --- | --- | --- |
| `lib/request/token-refresh.ts` | 147 | `shouldRefreshToken`, `refreshAndUpdateToken` + private refresh-retry predicates |
| `lib/request/error-classification.ts` | 366 | `isEntitlementError`, `isWorkspaceDisabledError`, `EntitlementError`, full unsupported-model fallback machinery |
| `lib/request/url-rewriting.ts` | 199 | `extractRequestUrl`, `rewriteUrlForCodex`, proxy resolution, shared dispatcher cache + cleanup |
| `lib/request/headers.ts` | 99 | `createCodexHeaders` (all overloads), RFC 8594 deprecation/sunset warning logging |
| `lib/request/fetch-helpers.ts` | 1,497 → 756 | Response/error orchestration and rate-limit parsing stay (they span concerns); explicit re-exports of every previously-public symbol |

- Re-exports are explicit lists (not `export *`), so the facade's public surface did not grow.
- Three previously-private helpers gained `export` in their new homes only, for cross-module use: `CHATGPT_CODEX_UNSUPPORTED_MODEL_CODE`, `isUnsupportedCodexModelForChatGpt`, `logDeprecationHeaders`.
- Runtime export surface compared before/after via namespace-import dump: byte-identical. No import cycles among the five request modules (madge).

## Validation

- [x] `npm run typecheck`
- [x] `npx eslint lib/request/ --max-warnings=0`
- [x] `npx vitest run` on all suites referencing fetch-helpers (`fetch-helpers`, `quota-probe`, `index`, `index-retry`, `public-api-contract`, `chaos/fault-injection`): 358/358 passed
- [x] Independently re-verified: typecheck + `fetch-helpers`/`public-api-contract` suites, 151/151

## Risk / Rollback

Mechanical move with a facade; revert the single commit to roll back. No data, config, or behavior changes.

https://claude.ai/code/session_01XNtnkLbBiXZxfQQYLMpucB

---
_Generated by [Claude Code](https://claude.ai/code/session_01XNtnkLbBiXZxfQQYLMpucB)_

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

splits `lib/request/fetch-helpers.ts` (1,497 lines) along natural seams into four focused modules (`token-refresh`, `error-classification`, `url-rewriting`, `headers`), keeping `fetch-helpers.ts` as an explicit re-export facade with zero behavior change. all code is moved verbatim; existing tests (358/358) pass unchanged.

- four new files extract token refresh, error classification, url/proxy logic, and header construction; each is standalone with no cross-module imports among themselves.
- facade re-export lists are explicit (not `export *`), preserving the exact public surface; three helpers gain `@internal` exports in sub-modules for use by `fetch-helpers.ts` itself, but are not forwarded through the facade.
- `registerCleanup(closeSharedProxyDispatchers)` is moved to `url-rewriting.ts` module-load time; behavior is unchanged since es module singletons guarantee it fires exactly once.

<h3>Confidence Score: 5/5</h3>

mechanical extraction with a re-export facade; all code is verbatim, public surface is byte-identical, and 358 tests pass — safe to merge.

every function, constant, and interface moved verbatim; no logic changes, no import cycles, no new behavior. the only new surface is three @internal-annotated helpers accessible via sub-path imports, which the previous thread already covers. the change is structurally clean.

no files need extra scrutiny; all five files carry straightforward extractions or re-export plumbing.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lib/request/error-classification.ts | new file; verbatim extraction of entitlement/unsupported-model logic from fetch-helpers.ts; three @internal exports (CHATGPT_CODEX_UNSUPPORTED_MODEL_CODE, isUnsupportedCodexModelForChatGpt) not re-exposed through facade — flagged in previous thread |
| lib/request/fetch-helpers.ts | trimmed to facade + response/rate-limit orchestration; explicit re-export lists correctly preserve every previously-public symbol |
| lib/request/headers.ts | new file; verbatim extraction of createCodexHeaders overloads + logDeprecationHeaders; logDeprecationHeaders exported @internal, not re-exposed through facade — flagged in previous thread |
| lib/request/token-refresh.ts | new file; verbatim extraction of shouldRefreshToken, refreshAndUpdateToken, and private retry predicates; EPERM/EBUSY retryability intact for windows token-file writes |
| lib/request/url-rewriting.ts | new file; verbatim extraction of proxy/URL rewrite logic; registerCleanup(closeSharedProxyDispatchers) side-effect preserved at module-load time; pre-existing double-create race in getSharedProxyDispatcher flagged in previous outside-diff comment |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    FH["fetch-helpers.ts<br/>(facade + response orchestration)"]
    EC["error-classification.ts<br/>entitlement · model fallback · workspace-disabled"]
    H["headers.ts<br/>createCodexHeaders · logDeprecationHeaders"]
    TR["token-refresh.ts<br/>shouldRefreshToken · refreshAndUpdateToken"]
    UR["url-rewriting.ts<br/>proxy resolution · URL rewrite<br/>+ registerCleanup side-effect"]

    EC -->|"re-exported via facade"| FH
    TR -->|"re-exported via facade"| FH
    UR -->|"re-exported via facade"| FH
    H  -->|"re-exported via facade"| FH

    FH -->|"internal import"| EC
    FH -->|"internal import"| H

    UR -->|"registerCleanup at load"| SD["shutdown.js"]
    TR -->|"queuedRefresh"| RQ["refresh-queue.js"]
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `lib/request/url-rewriting.ts`, line 1671-1680 ([link](https://github.com/ndycode/codex-multi-auth/blob/2d5dfa1fc3c8051ce1b9520a97cd0fe4bc2bae24/lib/request/url-rewriting.ts#L1671-L1680)) 

   <a href="#">P2</a> **concurrency: unguarded double-create in `getSharedProxyDispatcher`**

   two concurrent callers can both observe `sharedProxyDispatchers.get(proxyUrl) === undefined`, both construct a new `ProxyAgent`, and the first agent gets silently overwritten in the map — leaving an unclosed `ProxyAgent` that is never passed to `closeSharedProxyDispatchers`. this is pre-existing code moved verbatim, but isolating it in its own module makes it the single owner of `sharedProxyDispatchers`; no other module can close the leaked agent. on windows, an abandoned `ProxyAgent` can hold an open socket handle that blocks cleanup. consider checking-after-set or using a pending promise to serialize construction.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: lib/request/url-rewriting.ts
   Line: 1671-1680

   Comment:
   **concurrency: unguarded double-create in `getSharedProxyDispatcher`**

   two concurrent callers can both observe `sharedProxyDispatchers.get(proxyUrl) === undefined`, both construct a new `ProxyAgent`, and the first agent gets silently overwritten in the map — leaving an unclosed `ProxyAgent` that is never passed to `closeSharedProxyDispatchers`. this is pre-existing code moved verbatim, but isolating it in its own module makes it the single owner of `sharedProxyDispatchers`; no other module can close the leaked agent. on windows, an abandoned `ProxyAgent` can hold an open socket handle that blocks cleanup. consider checking-after-set or using a pending promise to serialize construction.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22ndycode%2Fcodex-multi-auth%22%20on%20the%20existing%20branch%20%22claude%2Faudit-06-fetch-helpers-split%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Faudit-06-fetch-helpers-split%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20lib%2Frequest%2Furl-rewriting.ts%0ALine%3A%201671-1680%0A%0AComment%3A%0A**concurrency%3A%20unguarded%20double-create%20in%20%60getSharedProxyDispatcher%60**%0A%0Atwo%20concurrent%20callers%20can%20both%20observe%20%60sharedProxyDispatchers.get%28proxyUrl%29%20%3D%3D%3D%20undefined%60%2C%20both%20construct%20a%20new%20%60ProxyAgent%60%2C%20and%20the%20first%20agent%20gets%20silently%20overwritten%20in%20the%20map%20%E2%80%94%20leaving%20an%20unclosed%20%60ProxyAgent%60%20that%20is%20never%20passed%20to%20%60closeSharedProxyDispatchers%60.%20this%20is%20pre-existing%20code%20moved%20verbatim%2C%20but%20isolating%20it%20in%20its%20own%20module%20makes%20it%20the%20single%20owner%20of%20%60sharedProxyDispatchers%60%3B%20no%20other%20module%20can%20close%20the%20leaked%20agent.%20on%20windows%2C%20an%20abandoned%20%60ProxyAgent%60%20can%20hold%20an%20open%20socket%20handle%20that%20blocks%20cleanup.%20consider%20checking-after-set%20or%20using%20a%20pending%20promise%20to%20serialize%20construction.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=ndycode%2Fcodex-multi-auth&pr=524&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a>
</details>

<a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22ndycode%2Fcodex-multi-auth%22%20on%20the%20existing%20branch%20%22claude%2Faudit-06-fetch-helpers-split%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Faudit-06-fetch-helpers-split%22.%0A%0AFix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Alib%2Frequest%2Ftoken-refresh.ts%3A1-10%0A**no%20dedicated%20vitest%20test%20file%20for%20the%20four%20new%20modules**%0A%0A%60token-refresh.ts%60%2C%20%60error-classification.ts%60%2C%20%60url-rewriting.ts%60%2C%20and%20%60headers.ts%60%20have%20no%20corresponding%20test%20files.%20existing%20coverage%20flows%20entirely%20through%20the%20%60fetch-helpers.ts%60%20facade%20tests%2C%20so%20per-module%20edge%20cases%20%28e.g.%20%60isRetryableAuthSetterError%60%20returning%20%60true%60%20on%20windows%20%60EPERM%60%2C%20the%20%60registerCleanup%60%20side-effect%20firing%20at%20module%20load%20in%20%60url-rewriting.ts%60%2C%20the%20%60%40internal%60-exported%20helpers%20being%20callable%20directly%20from%20sub-paths%29%20go%20untested%20in%20isolation.%20worth%20adding%20%60test%2Ftoken-refresh.test.ts%60%20etc.%20to%20lock%20down%20the%20extracted%20seams%20independently.%0A%0A&repo=ndycode%2Fcodex-multi-auth&pr=524&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
lib/request/token-refresh.ts:1-10
**no dedicated vitest test file for the four new modules**

`token-refresh.ts`, `error-classification.ts`, `url-rewriting.ts`, and `headers.ts` have no corresponding test files. existing coverage flows entirely through the `fetch-helpers.ts` facade tests, so per-module edge cases (e.g. `isRetryableAuthSetterError` returning `true` on windows `EPERM`, the `registerCleanup` side-effect firing at module load in `url-rewriting.ts`, the `@internal`-exported helpers being callable directly from sub-paths) go untested in isolation. worth adding `test/token-refresh.test.ts` etc. to lock down the extracted seams independently.


`````

</details>

<sub>Reviews (2): Last reviewed commit: ["docs(request): mark cross-module helpers..."](https://github.com/ndycode/codex-multi-auth/commit/52936722e4937ef11c889673a9160d48e3c7895a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36613959)</sub>

### CODE REVIEW

File: lib/request/error-classification.ts [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+		NORMALIZED_UNSUPPORTED_MODEL_PATTERN.test(bodyText) ||
+		MODEL_ACCESS_DENIED_PATTERN.test(bodyText)
+	);
+}
+
+export function getUnsupportedCodexModelInfo(
+	errorBody: unknown,
+): UnsupportedCodexModelInfo {
+	if (!isRecord(errorBody)) {
+		return { isUnsupported: false };
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/request/error-classification.ts:136]: Addressed in 5293672 with one deliberate deviation: re-exporting these three from the facade would *grow* the public surface (they were never importable before this PR), so each is instead marked `/** @internal Exported only for sibling lib/request modules; import via fetch-helpers.ts elsewhere. */`. The facade contract stays authoritative; nothing outside lib/request imports them by subpath.

---
_Generated by [Claude Code](https://claude.ai/code)_
